File: portal/backend/app/models/component.py
from sqlalchemy import Column, BigInteger, String, Integer, Text, DateTime, JSON
from sqlalchemy.sql import func

from app.core.database import Base


class ComponentHistory(Base):
    """组件历史版本快照"""
    __tablename__ = "component_history"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    component_id = Column(BigInteger, nullable=False, index=True, comment="关联组件 ID")
    version = Column(Integer, nullable=False, comment="快照时的版本号")
    name = Column(String(255), nullable=False)
    type = Column(String(50), nullable=False)
    description = Column(Text)
    config_json = Column(JSON, nullable=False, default=dict)
    params = Column(JSON, default=list)
    status = Column(String(50), nullable=False)
    created_by = Column(BigInteger)
    created_at = Column(DateTime, server_default=func.now())
    comment = Column(String(255), comment="版本备注")


class Component(Base):
    """统一组件模型 — sql / python / shell / datax"""
    __tablename__ = "component"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    name = Column(String(255), nullable=False)
    type = Column(String(50), nullable=False)  # sql / python / shell / datax
    description = Column(Text)
    # 配置 JSON,按 type 不同含义不同:
    #   sql:    { datasource_id, sql, timeout }
    #   python: { script, timeout }
    #   shell:  { script, timeout }
    #   datax:  { reader, writer, source_id, target_id, ... }
    config_json = Column(JSON, nullable=False, default=dict)
    # 参数定义 [{"key": "dt", "value": "${bizdate}", "desc": "业务日期"}]
    params = Column(JSON, default=list)
    version = Column(Integer, default=1, nullable=False)
    # 状态机: draft -> tested -> online -> offline
    status = Column(String(50), default="draft", nullable=False)
    # 发布后映射到 DS Task code (Phase 5+ 填入)
    ds_task_code = Column(BigInteger)
    folder_id = Column(BigInteger, comment="所属文件夹 id")
    sort_order = Column(Integer, default=0, nullable=False, comment="同文件夹内排序")
    previous_status = Column(String(50), comment="暂停前的状态")
    dqc_rule_ids = Column(JSON, nullable=True, comment="关联的数据质量规则 ID 列表")
    created_by = Column(BigInteger)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

File: portal/frontend/src/types/component.ts
export type Language = 'sql' | 'python' | 'shell' | 'datax'

export type ComponentStatus = 'draft' | 'tested' | 'online' | 'offline'

export interface ComponentParam {
  key: string
  value: string
  desc?: string
}

export interface ComponentItem {
  id: number
  name: string
  type: Language
  description?: string
  config_json: Record<string, any>
  params: ComponentParam[]
  version: number
  status: ComponentStatus
  status_label: string
  status_color: string
  ds_task_code?: number
  folder_id?: number
  sort_order: number
  code: string
  datasource_id?: number
  dqc_rule_ids?: number[]
  created_by?: number
  created_by_name?: string
  created_at?: string
  updated_at?: string
}

export interface FolderItem {
  id: number
  name: string
  type: string
  parent_id?: number | null
  sort_order: number
  depth?: number
}

export interface DatasourceItem {
  id: number
  name: string
  type: string
  host: string
  port: number
  database_name: string
  username: string
  description?: string
  status: number
  last_check_time?: string
  created_at?: string
}

export interface ProjectItem {
  id: number
  name: string
  description?: string
}

export interface StatusDef {
  label: string
  color: string
}

export const STATUS_DEFS: Record<ComponentStatus, StatusDef> = {
  draft:   { label: '草稿',   color: '#86909C' },
  tested:  { label: '已测试', color: '#0FC6C2' },
  online:  { label: '已上线', color: '#00B42A' },
  offline: { label: '已下线', color: '#FF7D00' },
}

export function statusLabel(s: string): string {
  return STATUS_DEFS[s as ComponentStatus]?.label || s
}

export function statusColor(s: string): string {
  return STATUS_DEFS[s as ComponentStatus]?.color || '#86909C'
}
